python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+        "xCol": {"doc": "input features column name"},
+        "outputCol": {"doc": "output column name"},
+        "yCol": {"doc": "optional label column name"},
+        "estimatorType": {"doc": "scikit-learn estimator type"},
+        "keyedSklearnEstimators": {"doc": "Dataframe of fitted sklearn estimators for each key"},
+        "outputType": {"doc": "SQL type for output column"}}
+    """`paramSpecs[paramName]` returns a parameter specification dictionary with the
+    documentation for the parameter of name `paramName`"""
+
+    _sql_types = {v: k for k, v in _type_mappings.items()}
+
+    @keyword_only
+    def __init__(self, keyCols=None, xCol=None, outputCol=None, yCol=None, estimatorType=None,
+                 keyedSklearnEstimators=None, outputType=None):
+        """Used by :class:`KeyedEstimator` to generate a :class:`KeyedModel`. Not intended for
+        external use."""
+
+        assert (estimatorType == "predictor") == (yCol is not None)
+        assert estimatorType == "transformer" or estimatorType == "predictor"
+        def implies(a, b): return not a or b
+        assert implies(estimatorType == "transformer", outputType == Vector.__UDT__)
+        assert len(keyCols) > 0
+        assert set(keyedSklearnEstimators.columns) == (set(keyCols) | set(["estimator"]))
+
+        for paramName, paramSpec in KeyedModel.paramSpecs.items():
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedModel, self).__init__()
+        if yCol and type(outputType) not in KeyedModel._sql_types:
+            raise TypeError("Output type {} is not an AtomicType (expected for {} estimator)"
+                            .format(outputType, estimatorType))
+        self._set(**self.__init__._input_kwargs)
+
+    def _transform(self, dataset):
+        keyCols = self.getOrDefault("keyCols")
+        xCol = self.getOrDefault("xCol")
+        outputCol = self.getOrDefault("outputCol")
+        outputType = self.getOrDefault("outputType")
+
+        _validateXCol(dataset.schema, xCol)
+
+        # Potential optimization: group input data by key, then only deserialize each
+        # estimator at most once. This becomes a bit difficult because then extraneous non-input
+        # columns must be preserved, yet they can't just be collected into single rows since
+        # we don't have a guarantee that they'll fit in memory.
+
+        # Potential optimization: broadcast estimator
+
+        shouldPredict = self.getOrDefault("estimatorType") == "predictor"
+        oneDimensional = _isOneDimensional(dataset.schema, xCol)
+        if shouldPredict:
+            cast = KeyedModel._sql_types[type(outputType)]
+        else:
+            assert outputType == Vector.__UDT__
+            # The closure of applyEstimator() doesn't know ahead of time that we won't use
+            # the cast value, so it tries to serialize it.
+            cast = None
```

Comment:
No, that'd remove it from the context.  I'm not trying to delete anything, but rather *give* it an explicit (though trivial value) so it *can* be serialized.